catalog/core/catalog-core-solr/src/main/java/ddf/catalog/source/solr/DynamicSchemaResolver.java

```diff
@@ -507,12 +507,18 @@ private Serializable getDocValue(String solrFieldName, Object docValue) {
    * @return the original field name
    */
   String resolveFieldName(String solrFieldName) {
-    int lastIndexOfUndercore = solrFieldName.lastIndexOf(FIRST_CHAR_OF_SUFFIX);
-
-    if (lastIndexOfUndercore != -1) {
-      return solrFieldName.substring(0, lastIndexOfUndercore);
+    String originalName = solrFieldName;
+    boolean suffixExists = true;
```

Comment:
:pencil2: See my comment about filtering out any field that is not part of the metacard type definition. I don't think you need these changes or the `SchemaFields.java` changes if you apply that change since metacard attributes should only contain `-` or `.` characters in their names. It looks like you only needed this change to filter out internal fields the Solr provider creates that uses only `_` as separators.  But those fields should never be returned from the provider.

Reply:
The original code only assumes that there is one suffix added to a field name in Solr, but that is actually not true.  For text fields we add one or more others depending on the field use (tokenization, case sensitivity, phonetics).  From a highlighting perspective, I can get a hit on any of those other fields.  For example, let's say I get a highlight on the tokenized version of the `title` field.  The name of that field in Solr is `title_txt_tokenized`.  A call to this function in its original form would return `title_txt`, which is not the real, original field name.  Even worse, let's assume I get a highlight hit against the case sensitive version of the title field, which in Solr is `title_txt_tokenized_has_case`.  In this example I get returned the value `title_txt_tokenized_has`.  In the original use case of this function, it was pretty safe to assume that the `_` character marked the suffix because it was really only ever called when a field only had one suffix (the assumed stored field for returning results).  Since highlights can occur against these other fields, I do need these changes.  Plus, this removes the assumptions about the field names and makes it more consistent with how the resolver actually generates all of the Solr field names.

Reply:
We would have to benchmark it to be sure either way so this might be premature optimization but my initial concern was that this method is called on every field of every metacard converted back from a Solr doc.

Another option would be not to care if the suffix is at the end.  Then you could pass over the suffix list once.  A slight optimization for that would be to find the index of the first underscore and use that as an offset when getting the `indexOf` of the suffix since we know all suffixes start with an underscore.

Reply:
I mostly reverted this code but changed it so that it looks for the first index rather than the last.  Since this should only be called with real attribute names, it should take care of all of the permutations of the added suffixes.